### ISSUE TITLE
where filter fails on comparing numeric values to strings with nested objects

### ISSUE THREAD

**Issue by mahmoudmoravej** [Shopify/liquid#1709]
I have a nested object structure comes from here: https://shopify.dev/docs/api/liquid/objects/selling_plan_allocation

all the followings commands return **empty**:
```liquid
variant.selling_plan_allocations | map: 'selling_plan'  | where: 'id', '123' | first
```
```liquid
variant.selling_plan_allocations | where: 'selling_plan.id', '123' | first
```
```liquid
variant.selling_plan_allocations | where: 'selling_plan.id', 123 | first
```

and this is the only thing **works**:
```liquid
variant.selling_plan_allocations | map: 'selling_plan'  | where: 'id', 123 | first
```

**Comment by mahmoudmoravej** [comment]
You're right but as I mentioned the 3rd command is not working as well. I mean:
```
variant.selling_plan_allocations | where: 'selling_plan.id', 123 | first
```
is not working although I passed numeric one.

**Comment by moritz-pv** [comment]
That's because 'selling_plan.id' is not a key of a property belonging to a selling_plan_allocation. It's a key of a nested property. I know this is exactly what you claim to be a bug, but I think it is not a bug. It is just poor language design that everything is restricted like that and one has to find cumbersome workarounds with extra variables and loops for almost everything in this stupid language.
In this poor excuse of a language it is not even possible to properly evaluate most boolean expressions within if-statements, because there is no concept of operator precedence, it is just "left to right" and parentheses are not allowed. So "a and b or c and d" would be treated as "((a and b) or c) and d". It is possible to find a very cumbersome workaround for that by storing extra intermediate variables (adding multiple confusing extra if-statements to do that, of course, because one can't simply assign a boolean expression as a variable) and then to continue from that.
Similarly you could find an elaborate and cumbersome workaround to solve your problem in maybe 20-30 lines of code.

TLDR: Liquid is designed to be cumbersome and awkward to use for everything but the most basic control flows. If you get frustrated and desperate trying to do anything more "advanced" that is probably intended.

**Comment by moritz-pv** [comment]
> Similarly you could find an elaborate and cumbersome workaround to solve your problem in maybe 20-30 lines of code

You even only need six lines:
```
for a in variant.selling_plan_allocations
  if v.selling_plan.id == 123
    assign result = a
    break
  endif
endfor
```
Really, what better can you expect from a language designed by a dyslexic person?

Compare that to the "a and b or c and d" problem:
```
if a and b
  assign a_and_b = true
else
  assign a_and_b = false
endif
if c and d
  assign c_and_d = true
else
  assign c_and_d = false
endif
if a_and_b or c_and_d
  assign a_and_b_or_c_and_d = true
else
  assign a_and_b_or_c_and_d = false
endif
```
